### PR TITLE
[Optimize] Move memo() from BlockStyles to BlockPreview

### PR DESCRIPTION
**Pull request by adamziel**
## Description
https://github.com/WordPress/gutenberg/pull/21973 optimized `<BlockStyles />` to avoid unnecessary re-renders, specifically it replaced HOC with hooks, and added a `memo()` around the component. As discussed in https://github.com/WordPress/gutenberg/pull/21990#issuecomment-621793896, `<BlockPreview />` seems to be a much better candidate for `memo()` - especially since it's reused in other places. This PR moves the `memo()`, while keeping the performance boost we got in #21973.

## How has this been tested?
Tested locally with the same test plan as in #21973

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by adamziel** [comment]
It seems like replacing `export default BlockPreview;` with `export default memo( BlockPreview );` breaks the README which now says the component is undocumented. What's the best way around that?

**Comment by youknowriad** [comment]
> It seems like replacing export default BlockPreview; with export default memo( BlockPreview ); breaks the README which now says the component is undocumented. What's the best way around that?

yes you can probably do something like

```
/**
* docs
*/
const BlockPreview = memo(props => ())
```

File: packages/block-editor/src/components/block-styles/index.js [hunk 1]
```diff
@@ -127,67 +122,35 @@ function BlockStyles( { clientId, onSwitch = noop, onHoverClassName = noop } ) {
...
 	block,
 	blockName,
-	styleClassName,
-	viewportWidth,
+	useExample,
+	style,
+	activeStyle,
+	updateClassName,
+	className,
+	onHoverClassName,
```
Comment (youknowriad) [inline packages/block-editor/src/components/block-styles/index.js:150]: Some prop suggestions:

- activeStyle can be replaced with isActive
- `onHoverClassName` should we rename this `onHover` and accept  the style name (or object as  argument instead of the className.
- Can we rename `updateClassname` as `onSelect`, I also think the argument is not necessary since the parent  component already knows which style is being rendered?
-  It also  seems like instead of  `useExample`, `type ` and `block` we can just compute a base "block" on the p arent and pass just that block. (Unless I'm missing something, the base block  is always the same, we just alter its  className attribute per style)

**Review by youknowriad — APPROVED** [review]

**Comment by adamziel** [comment]
@youknowriad I addressed your feedback, but since we're cleaning it up I don't like how `styleClassName` is now passed to `BlockStyleItem` just to compute the specific version of the `genericPreviewBlock`. I played with computing everything at the top level in https://github.com/WordPress/gutenberg/pull/22133 - I'd really appreciate your feedback on both PRs

**Comment by youknowriad** [comment]
> I don't like how styleClassName is now passed to BlockStyleItem just to compute the specific version of the genericPreviewBlock

In my mind, I'd expect `BlockStyleItem` to just receive the `style` object and compute `styleClassName` itself,  it is made a bit complex by the fact that you need to know the active style to remove it first, so this makes me wonder wheter the "genericBlock" should come without any style variations applied and `BlockStyleItem` would just add its own class. (the current code doesn't bother too much though)

I  took a quick look at the second PR and  IMO, I feel we lose a bit of clarity in the code, I found it hard to follow but maybe it's just me. I'd be ok with it if it's a big performance improvement but personally, I prefer the current PR.

**Comment by adamziel** [comment]
>  it is made a bit complex by the fact that you need to know the active style to remove it first,

Also by the fact the parent needs to know that class name in order to update block attributes and call `onHoverClassName`. It would be great to improve this part at some point, but for now I'll leave it as it is.
